File: app/src/main/java/com/middlename/budgetingapp/MainActivity.java
package com.middlename.budgetingapp;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.middlename.budgetingapp.databinding.ActivityMainBinding;

import java.util.List;

public class MainActivity extends AppCompatActivity implements ClickEvent {
    ActivityMainBinding binding;
    TransactionsAdapter transactionsAdapter;
    TransactionsDatabase transactionsDatabase;
    TransactionsDao transactionsDao;
    long balance = 0, income = 0, expense = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, AddTransaction.class));
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();

        balance = 0;
        income = 0;
        expense = 0;

        transactionsDatabase = TransactionsDatabase.getInstance(this);
        transactionsDao = transactionsDatabase.getDao();
        transactionsAdapter = new TransactionsAdapter(this, this);
        binding.transactions.setAdapter(transactionsAdapter);
        binding.transactions.setAdapter(transactionsAdapter);
        binding.transactions.setLayoutManager(new LinearLayoutManager(this));

        List<TransactionsTable> transactionsTables = transactionsDao.getAll();

        for (TransactionsTable transactionsTable : transactionsTables) {
            if (transactionsTable.isIncome()) { income += transactionsTable.getAmount(); }
            else { expense += transactionsTable.getAmount(); }

            transactionsAdapter.add(transactionsTable);
        }
        balance = income - expense;

        binding.incomeAmount.setText(income + "");
        binding.expensesAmount.setText(expense + "");
        binding.balanceAmount.setText(balance + "");
    }

    @Override
    public void OnClick(int position) {
        Intent intent = new Intent(MainActivity.this, AddTransaction.class);
        intent.putExtra("update", true);
        intent.putExtra("id", transactionsAdapter.getTransactionId(position));
        intent.putExtra("name", transactionsAdapter.getTransactionName(position));
        intent.putExtra("amount", transactionsAdapter.getTransactionAmount(position));
        intent.putExtra("description", transactionsAdapter.getTransactionDescription(position));
        intent.putExtra("date", transactionsAdapter.getTransactionDate(position));
        intent.putExtra("type", transactionsAdapter.getTransactionType(position));
        startActivity(intent);
    }
}

File: app/src/main/java/com/middlename/budgetingapp/MainActivity.kt
package com.middlename.budgetingapp

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity() {

    private lateinit var toolbar: Toolbar
    private lateinit var balanceTextView: TextView
    private lateinit var addTransactionFab: FloatingActionButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        toolbar = findViewById(R.id.toolbar)
        balanceTextView = findViewById(R.id.current_balance_text_view)
        addTransactionFab = findViewById(R.id.add_transaction_fab)

        setSupportActionBar(toolbar)
//        supportActionBar?.setDisplayHomeAsUpEnabled(true)
//        supportActionBar?.setDisplayShowHomeEnabled(true)

        addTransactionFab.setOnClickListener {
//            code
        }

        // Update balance text view with the current balance
        val currentBalance = getCurrentBalance()
        balanceTextView.text = getString(R.string.current_balance_template, currentBalance)

        showSampleData()
    }

    private fun getCurrentBalance(): Double {
        // TODO: Implement method to get the current balance
        return 0.0
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

//    show sample data to recycler view
    private fun showSampleData() {
        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = TransactionAdapter(getSampleData())
    }

    private fun getSampleData(): List<Transaction> {
        return listOf(
            Transaction("Sample Transaction 1", 100.0),
            Transaction("Sample Transaction 2", 200.0),
            Transaction("Sample Transaction 3", 300.0),
            Transaction("Sample Transaction 4", 400.0),
            Transaction("Sample Transaction 5", 500.0),
            Transaction("Sample Transaction 6", 600.0),
            Transaction("Sample Transaction 7", 700.0),
            Transaction("Sample Transaction 8", 800.0),
            Transaction("Sample Transaction 9", 900.0),
            Transaction("Sample Transaction 10", 1000.0),
            Transaction("Sample Transaction 11", 1100.0),
            Transaction("Sample Transaction 12", 1200.0),
            Transaction("Sample Transaction 13", 1300.0),
            Transaction("Sample Transaction 14", 1400.0),
            Transaction("Sample Transaction 15", 1500.0),
            Transaction("Sample Transaction 16", 1600.0),
            Transaction("Sample Transaction 17", 1700.0),
            Transaction("Sample Transaction 18", 1800.0),
            Transaction("Sample Transaction 19", 1900.0),
            Transaction("Sample Transaction 20", 2000.0),
            Transaction("Sample Transaction 21", 2100.0),
            Transaction("Sample Transaction 22", 2200.0),
            Transaction("Sample Transaction 23", 2300.0),
            Transaction("Sample Transaction 24", 2400.0),
            Transaction("Sample Transaction 25", 2500.0),
            Transaction("Sample Transaction 26", 2600.0),
            Transaction("Sample Transaction 27", 2700.0),
            Transaction("Sample Transaction 28", 2800.0),
            Transaction("Sample Transaction 29", 2900.0),
            Transaction("Sample Transaction 30", 3000.0)
        )
    }
}

